### ISSUE TITLE
request.stream.read() doesnt receive any data from simulate_post

### ISSUE THREAD

**Issue by b3g00d** [falconry/falcon#1565]
I have example code in a test file

```
def test_group(self):                                              
    result = self.simulate_post(                                   
        path="/threads/group",                                     
        headers={"Authorization": "Bearer {}".format(self.token)}, 
        json={"user_ids": ["1234", "4567"], "name": "test"})
    self.assertEqual(result.status_code, 200)                      
```

and a handler for it is

```
def on_post(self, req, resp):                    
    user_id = self.user_id                       
    try:                                         
        body = json.loads(req.stream.read())     
    except Exception:                            
        resp.status = falcon.HTTP_403            
        resp.media = {"message": "Body required"}
        return                                   
    if not body:                                 
        resp.status = falcon.HTTP_403            
        resp.media = {"message": "Body required"}
        return                                   
```
And it keep producing AssertionError because of req.stream.read().
I already test with requests lib and handler working fine.
What did I do wrong in the test file?

**Comment by vytas7** [comment]
Hi @b3g00d !
I can indeed reproduce your issue.

However, [`req.stream`](https://falcon.readthedocs.io/en/stable/api/request_and_response.html#falcon.Request.stream) behaviour, as detailed in the docs (linked), is very dependent upon the WSGI server implementation. When emulating requests with the Falcon testing framework, the server is effectively Python standard [`wsgiref`](https://docs.python.org/3.7/library/wsgiref.html), with some Falcon wrappers around.

If you are concerned with the `req.stream` behaviour and don't might a slight performance penalty, [`req.bounded_stream`](https://falcon.readthedocs.io/en/stable/api/request_and_response.html#falcon.Request.bounded_stream) could be used to normalize the behaviour across WSGI servers.
Indeed, the following passes for me:
```python
import json

import falcon
import falcon.testing


class Resource:

    def on_post(self, req, resp):
        try:
            body = json.loads(req.bounded_stream.read().decode())
        except Exception:
            resp.status = falcon.HTTP_403
            resp.media = {"message": "Body required"}
            return

        if not body:
            resp.status = falcon.HTTP_403
            resp.media = {"message": "Body required"}
            return

        print("Got body: {!r}".format(body))


class TestCase(falcon.testing.TestCase):

    def setUp(self):
        super().setUp()

        self.token = "very_secure"

        api = falcon.API()
        api.add_route('/threads/group', Resource())
        self.app = api

    def test_post(self):
        result = self.simulate_post(
            path="/threads/group",
            headers={"Authorization": "Bearer {}".format(self.token)},
            json={"user_ids": ["1234", "4567"], "name": "test"})

        self.assertEqual(result.status_code, 200)
```

The code is heavily based on yours, but `req.stream.read()` has been changed to `req.bounded_stream.read()`.

An even easier course of action is using [Falcon's media facilities](https://falcon.readthedocs.io/en/stable/api/media.html): `req.media` would understand your JSON payload directly:
```python
import json

import falcon
import falcon.testing


class Resource:

    def on_post(self, req, resp):
        print("Got media: {!r}".format(req.media))


class TestCase(falcon.testing.TestCase):

    def setUp(self):
        super().setUp()

        self.token = "very_secure"

        api = falcon.API()
        api.add_route('/threads/group', Resource())
        self.app = api

    def test_post(self):
        result = self.simulate_post(
            path="/threads/group",
            headers={"Authorization": "Bearer {}".format(self.token)},
            json={"user_ids": ["1234", "4567"], "name": "test"})

        self.assertEqual(result.status_code, 200)
```

Hope this helps, but don't hesitate to ask if this still doesn't answer your case!

**Comment by b3g00d** [comment]
@vytas7 thanks for the help.

**Comment by jeffdico** [comment]
thanks for this information. I was having the same issue and my code read req.stream.read(). 
Now i have changed it to req.bound_stream.read() and it works well. Thanks alot.

**Comment by onurgosteris** [comment]
> 

req.bound_stream.read() instead req.bounded_stream.read()

Thank you so much

**Comment by kentbull** [comment]
This is still happening for me.

**Comment by CaselIT** [comment]
@kentbull please open a new issue or discussion, thanks

**Comment by MorgusLethe** [comment]
I came here from google. ChatGPT wrote some code for me that used request.stream.read()

Had issues where I was unable to POST - preflight succeeded, but the POST did not. The problem was that the POST used  request.stream.read() which froze up the code. Strangely, refreshing the page made the reqest succeed, which confused me even more. I spent hours thinking it was CORS (that's what the browser console was telling me), but in the end, when I read this issue, I used req.media instead of request.stream.read() and now it's working.

I used wsgiref.

Just dropping this note in case it might help someone else search this up.
